### ISSUE TITLE
Expose Pod cgroup path as part of `status` field

### ISSUE THREAD

**Issue by aditighag** [kubernetes/kubernetes#113342]
### What would you like to be added?

We would like pod's cgroup path to be exposed in the Pod `status` (or other metadata field in the Pod object).  

### Why is this needed?

This would allow accessing pod's cgroup paths in a way that is independent of the container runtime, or cgroup driver, etc, and that doesn't require installing additional tooling. 

Here are some use cases utilizing pod cgroup paths:

- Cilium uses pod cgroup paths to identify pods for eBPF datapath events related to observability, load-balancing, etc. 

- Per-cgroup [BPF LSM](https://docs.kernel.org/bpf/prog_lsm.html) programs to implement security policies.

- Ability to install BPF cgroup programs on a per pod basis based on cgroup hierarchies to control how network traffic is processed.

- There are other open source projects like [DataDog](https://github.com/DataDog/datadog-agent/blob/67ce06c03e207193ac6d9f38481755e323cdb1f6/pkg/util/cgroups/reader.go#L123), [Pixie](https://github.com/pixie-io/pixie/blob/98189f6e939a9e7787ab31feca7ce3f7633a44ed/src/shared/metadata/cgroup_path_resolver.cc#L79), [Inspektor gadget](https://github.com/inspektor-gadget/inspektor-gadget/blob/a219c9c29c67d07e41b36bd762d8330915f554db/pkg/container-utils/cgroups/cgroups.go#L99) with use cases such as for monitoring, observability, events/stats collection, etc. 

**Current limitations with retrieving pod's cgroup path:**

Following are the possible ways of getting this information, all with their own limitations. 

- Tools such as [crictl](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md): It's a CLI tool for container runtimes that runs on k8s nodes.
This is an example of how you can retrieve the cgroup path for a pod:
 
```
$ sudo crictl inspectp -o=json $POD_ID | grep cgroupsPath
"cgroupsPath": "/kubepods/burtsable/podd7f4b509-cf94-4951-9417-d1087c92a5b2/7ccf55aee35dd16aca4189c952d83487297f3cd760f1bbf09620e206e7d0c27a"
```

Cons: The tool may not be installed on minimalistic OS's, and the API isn't guaranteed to be stable.

- CRI gRPC interface: We could connect to container runtimes over the CRI.

Cons: This approach requires root privileges, and hence has security implications. 

- Crafting based on patterns: Pods' hierarchical cgroup paths are built based on known default patterns. 

Cons: 
1) No standard way to construct paths, and they can vary based on a variety of variable factors such as pod QoS, pod ID,  cgroup driver, container runtime, nested environments like KIND, etc. This approach is quite cumbersome and not reliable to support different environments: here is an example of a complex [logic](https://github.com/pixie-io/pixie/blob/98189f6e939a9e7787ab31feca7ce3f7633a44ed/src/shared/metadata/cgroup_path_resolver.cc#L79).
3) Every project needs to come up with their own (duplicated) ways to construct pod cgroup paths.

Sample container's cgroup path with systemd cgroup driver:

```
kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-besteffort.slice/kubelet-kubepods-besteffort-pod9ac48755_3968_48e4_b9dc_6d4b69f3bb42.slice/cri-containerd-3baf66ee56a52a8765c3deb2444315411a888fa3e2f8f7ddd75e9ded3c34425e.scope
```

- Parsing /proc on host: Parse `/proc/$PID/cgroup` on underlying host. 

Cons: 
1) Need additional bookkepping of pod/container pids.
2) Security implications with mounting `/proc` inside pods/containers tasked with these operations.

In summary, a standardized way of getting pods' cgroup paths as part of Kubernetes API would help the community to simplify, and support a variety of use cases.

**Comment by n4j** [comment]
@aditighag This looks like a new feature request which would require consensus from the SIG Node steering committee.

You would need to present your case to one of the SIG Node [meetings](https://github.com/kubernetes/community/tree/master/sig-node) based on the outcome this could be a new [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/README.md)

**Comment by n4j** [comment]
@aditighag Yes, add it to the agenda

**Comment by n4j** [comment]
@aditighag Yes, add it to the agenda

**Comment by kad** [comment]
Few comments:
- cgroup is Linux specific. exposing that info to Pod status on API server, might be not portable between OSes.
-  VM based runtimes: host cgroup might not be the same as inside VM.
- libraryos based runtimes: cgroups might not be really in use.

idea/suggestion: kubelet's PodResources local endpoint?

**Comment by mrunalp** [comment]
This cgroup path could be passed at CNI setup time and the plugin could cache it (if we want to punt on status change).

**Comment by dchen1107** [comment]
> idea/suggestion: kubelet's PodResources local endpoint?

Was considered before, but have the security concerns raised by SIG Auth.

**Comment by mikebrow** [comment]
@MikeZappa87 heads up we also need to add the pod's cgroup path to the json we pass to the cni plugins... @haircommander @bobbypage

**Comment by MikeZappa87** [comment]
@mikebrow would this be via CNI_ARG/annotation/runtime config?

**Comment by squeed** [comment]
@aditighag and I are talking about injecting this as a cni ["capability arg"](https://github.com/containernetworking/cni/blob/main/CONVENTIONS.md#well-known-capabilities)

**Comment by MikeZappa87** [comment]
> @aditighag and I are talking about injecting this as a cni ["capability arg"](https://github.com/containernetworking/cni/blob/main/CONVENTIONS.md#well-known-capabilities)

This should be pretty cut/dry. Sandbox_run.go and then the namespace_opts in go-cni. I can help move this along

**Comment by aditighag** [comment]
@MikeZappa87 Great, yes, please help with the containerd/go-cni changes. Thanks!
Here is the PR to extend capabilities - https://github.com/containernetworking/cni/pull/936. /cc @caseydavenport

**Comment by aditighag** [comment]
> @MikeZappa87 heads up we also need to add the pod's cgroup path to the json we pass to the cni plugins... @haircommander @bobbypage

What's the plan for the container runtime related changes?
@MikeZappa87 is working on the containerd/go-cni changes. 
@haircommander Anything else needed besides the capabilities extension [PR](https://github.com/containernetworking/cni/pull/936) before you can get started on cri-o front?

**Comment by haircommander** [comment]
We will have to add a patch to github.com/cri-o/ocicni, which can be done once the aforementioned PR merges

**Comment by aditighag** [comment]
The [cgroup path capability](https://github.com/containernetworking/cni/pull/936) was released in containerd 1.7.1, and the cri-o change was merged on May 10 (there hasn't been any new release since then). /cc @squeed

**Comment by aditighag** [comment]
@n4j / @squeed I don't have the permissions to edit the labels, but you can update them accordingly. 

/fixed

**Comment by emzxcv** [comment]
is there any chance of it being planned again?
